ruma-events-macros/src/event_enum.rs

```diff
@@ -0,0 +1,275 @@
+                            Ok(#ident::#variants(event))
+                        },
+                    )*
+                    _ => Err(D::Error::custom(format!("event type `{}` is not a valid event", ev_type)))
+                }
+            }
+        }
+    };
+
+    let event_content_enum = expand_content_enum(input)?;
+
+    Ok(quote! {
+        #event_enum
+
+        #event_deserialize_impl
+
+        #event_content_enum
+    })
+}
+
+/// Create a content enum from `EventEnumInput`.
+pub fn expand_content_enum(input: EventEnumInput) -> syn::Result<TokenStream> {
+    let attrs = &input.attrs;
+    let ident = Ident::new(&format!("{}Content", input.name.to_string()), input.name.span());
+    let event_type_str = &input.events;
+
+    let variants = input.events.iter().map(to_camel_case).collect::<syn::Result<Vec<_>>>()?;
+    let content = input.events.iter().map(to_event_content_path).collect::<Vec<_>>();
+
+    let content_enum = quote! {
+        #( #attrs )*
+        #[derive(Clone, Debug, ::serde::Serialize)]
+        #[serde(untagged)]
+        #[allow(clippy::large_enum_variant)]
+        pub enum #ident {
+            #(
+                #[doc = #event_type_str]
+                #variants(#content),
+            )*
+            /// Any custom event.
+            Custom(::ruma_events::custom::CustomEventContent),
+        }
+    };
+
+    let event_content_impl = quote! {
+        impl ::ruma_events::EventContent for #ident {
+            fn event_type(&self) -> &str {
+                match self {
+                    #( Self::#variants(content) => content.event_type(), )*
+                    Self::Custom(content) => content.event_type(),
+                }
+            }
+
+            fn from_parts(event_type: &str, input: Box<::serde_json::value::RawValue>) -> Result<Self, ::serde_json::Error> {
+                match event_type {
+                    #(
+                        #event_type_str => {
+                            let content = #content::from_parts(event_type, input)?;
+                            Ok(#ident::#variants(content))
+                        },
+                    )*
+                    ev_type => {
+                        let content = ::ruma_events::custom::CustomEventContent::from_parts(ev_type, input)?;
+                        Ok(#ident::Custom(content))
+                    },
+                }
+            }
+        }
+    };
+
+    let any_event_variant_impl = quote! {
+        impl #ident {
+            fn is_compatible(event_type: &str) -> bool {
+                match event_type {
+                    #( #event_type_str => true, )*
+                    _ => false,
+                }
+            }
+        }
+    };
+
+    let marker_trait_impls = marker_traits(&ident);
+
+    Ok(quote! {
+        #content_enum
+
+        #any_event_variant_impl
+
+        #event_content_impl
+
+        #marker_trait_impls
+    })
+}
+
+fn marker_traits(ident: &Ident) -> TokenStream {
```

Comment:
This function seems to be defined twice, once here in `event_enum` and once in `event_content`.

Reply:
Also.. Do we need this? With this design, we don't even really want the user to be able to create a `StateEvent<AnyStateEventContent>`, why would the generated types implement the event content marker traits?